ciongke/cmd/hook/main.go

```diff
@@ -0,0 +1,352 @@
+		return
+	}
+	sig := r.Header.Get("X-Hub-Signature")
+	if sig == "" {
+		http.Error(w, "403 Forbidden: Missing X-Hub-Signature", http.StatusForbidden)
+		return
+	}
+
+	// Validate the payload with our HMAC secret.
+	payload, err := github.ValidatePayload(r, s.HMACSecret)
+	if err != nil {
+		http.Error(w, "403 Forbidden: Invalid X-Hub-Signature", http.StatusForbidden)
+		return
+	}
+	fmt.Fprintf(w, "Event received. Have a nice day.")
+	go func() {
+		s.Events <- Event{eventType, payload}
+	}()
+}
+
+// handleEvent unmarshals the payload and dispatches to the correct handler.
+func (s *Server) handleEvent(e Event) {
+	// TODO: "test this" comment
+	switch e.Type {
+	case "pull_request":
+		var pr github.PullRequestEvent
+		if err := json.Unmarshal(e.Payload, &pr); err != nil {
+			log.Printf("Could not unmarshal pull request event payload: %s", err)
+			return
+		}
+		if err := s.handlePullRequestEvent(pr); err != nil {
+			log.Printf("Error handling pull request: %s", err)
+			return
+		}
+	}
+}
+
+// handlePullRequestEvent decides what to do with PullRequestEvents.
+func (s *Server) handlePullRequestEvent(pr github.PullRequestEvent) error {
+	switch pr.Action {
+	case "opened", "reopened", "synchronize":
+		valid, err := s.validAuthor(pr.PullRequest.User.Login)
+		if err != nil {
+			return fmt.Errorf("could not validate author: %s", err)
+		}
+		if valid {
+			if err := s.buildPR(pr.PullRequest); err != nil {
+				return fmt.Errorf("could not build PR: %s", err)
+			}
+		} else if pr.Action == "opened" {
+			// TODO: Comment asking them to join the team.
+		}
+	case "closed":
+		if err := s.deletePR(pr.PullRequest.Base.Repo.Name, pr.Number); err != nil {
+			return fmt.Errorf("could not delete old PR: %s", err)
+		}
+	}
+	return nil
+}
+
+// validAuthor first checks if author is in the org, then the team.
+func (s *Server) validAuthor(author string) (bool, error) {
+	orgMember, err := s.GitHubClient.IsMember(s.Org, author)
+	if err != nil {
+		return false, err
+	}
+	if orgMember {
+		return true, nil
+	}
+	teamMember, err := s.GitHubClient.IsTeamMember(s.TeamID, author)
+	if err != nil {
+		return false, err
+	}
+	return teamMember, nil
+}
+
+// buildPR deletes any jobs building the PR and then starts a new one.
+func (s *Server) buildPR(pr github.PullRequest) error {
+	name := fmt.Sprintf("%s-pr-%d-%s", pr.Base.Repo.Name, pr.Number, pr.Head.SHA[:8])
+	job := &batch.Job{
+		ObjectMeta: api.ObjectMeta{
+			Name:      name,
+			Namespace: s.Namespace,
+			Labels: map[string]string{
+				"repo": pr.Base.Repo.Name,
+				"pr":   strconv.Itoa(pr.Number),
+			},
+		},
+		Spec: batch.JobSpec{
+			ActiveDeadlineSeconds: &jobDeadlineSeconds,
+			Template: api.PodTemplateSpec{
+				Spec: api.PodSpec{
+					RestartPolicy: api.RestartPolicyNever,
+					Containers: []api.Container{
+						{
+							Name:  "test-pr",
+							Image: s.TestJobImage,
+							Command: []string{
```

Comment:
Is there a reason we need to use `Command` rather than `args`? Letting the image choose the `ENTRYPOINT` seems like better coupling to me.

http://kubernetes.io/docs/api-reference/v1/definitions/#_v1_container